src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -156,72 +134,35 @@ internal static bool ReleaseJSObject(JSObject objToRelease)
             return true;
         }
 
-        public static void UnBindRawJSObjectAndFree(int gcHandle)
-        {
-            GCHandle h = (GCHandle)(IntPtr)gcHandle;
-            JSObject? obj = h.Target as JSObject;
-            lock (_rawToJS)
-            {
-                if (obj?.RawObject != null)
-                {
-                    _rawToJS.Remove(obj.RawObject);
-                    obj.FreeHandle();
-                }
-            }
-        }
-
-        public static object CreateTaskSource(int jsId)
+        public static int CreateTaskSource()
         {
-            return new TaskCompletionSource<object>();
+            var tcs= new TaskCompletionSource<object>();
+            return GetJSOwnedObjectHandle(tcs);
         }
 
-        public static void SetTaskSourceResult(TaskCompletionSource<object> tcs, object result)
+        public static void SetTaskSourceResult(int tcsGCHhanle, object result)
         {
+            GCHandle handle = (GCHandle)(IntPtr)tcsGCHhanle;
+            TaskCompletionSource<object> tcs = (TaskCompletionSource<object>)handle.Target!;
```

Comment:
I think the ```Target!``` in these functions is incorrect, the handle could theoretically be weak

Reply:
No, it could not be weak. 
This is JS owned Normal handle. We always have a Target.
Added comment

Reply:
I mean, event if we implement weak listeners on the JS side, it would be just weak reference to the js function.
The js function would still hold strong ref to the C# side.